### PR TITLE
Only ensure solutions are in the same file in cargo fix

### PR DESCRIPTION
**Pull request by pietroalbini**
This PR changes `cargo fix` to avoid rejecting machine-applicable lints with multiple suggestions. Instead, now it only checks the solutions are in the same file.

I don't know how to write a test for this, since no lint in rustc relies on the new behavior and the existing `cargo fix` tests just invoke lints in rustc. Any idea on that would be appreciated.

r? @alexcrichton 
cc @killercup 
fixes https://github.com/rust-lang/cargo/issues/6401

### CODE REVIEW

**Comment by pietroalbini** [comment]
Trying this with my local rustc branch doesn't work yet, let me fix it.

**Comment by pietroalbini** [comment]
Nevermind, ran the wrong cargo binary... This patch works fine.

it works!

**Comment by pietroalbini** [comment]
I'd like to also include a bump of rustfix in this PR, once https://github.com/rust-lang-nursery/rustfix/pull/155 is merged.

File: src/cargo/ops/fix.rs [hunk 1]
```diff
@@ -400,21 +400,21 @@ fn rustfix_and_fix(
     for suggestion in suggestions {
         trace!("suggestion");
         // Make sure we've got a file associated with this suggestion and all
-        // snippets point to the same location. Right now it's not clear what
-        // we would do with multiple locations.
-        let (file_name, range) = match suggestion.snippets.get(0) {
-            Some(s) => (s.file_name.clone(), s.line_range),
+        // snippets point to the same file. Right now it's not clear what
+        // we would do with multiple files.
+        let file_name = match suggestion.solutions.get(0).and_then(|sol| sol.replacements.get(0)) {
```
Comment (killercup) [inline src/cargo/ops/fix.rs]: Not sure if this is the right place to make such a change, but I'd try to extract that "file name getter":

```rust
let file_names = suggestion.solutions.iter()
    .flat_map(|s| s.replacements.iter())
    .map(|r| &r.snippet.file_name);

let file_name = if let Some(file_name) = file_names.clone().next() {
    file_name.clone()
} else {
    trace!("rejecting as it has no solutions {:?}", suggestion);
    continue;
};

if !file_names.clone().all(|f| f == &file_name) {
    trace!("rejecting as it changes multiple files: {:?}", suggestion);
    continue;
}
```

**Comment by pietroalbini** [comment]
Ok, removed the bump to rustfix 0.4.3. This is ready to be reviewed, even though it's not enough on its own for https://github.com/rust-lang/rust/pull/56645 (needs a fix for https://github.com/rust-lang/rust/issues/53934 before I can fix all the `unused_import`s).

**Pull request merged**